Sprint#17 Homework

ekalemen opened the pull request:
Use dependancy injection framework Koin to make dependasies more flexible and get rid of Creator class.

Commit:
Sprint#17 Homework

Use dependancy injection framework Koin to make dependasies more
flexible and get rid of Creator class.

Signed-off-by: Eugeny Kalemenev <e.kalemenev@yadro.com>

ArturNurtdinov commented on app/build.gradle.kts:
```diff
@@ -50,6 +50,7 @@ dependencies {
     implementation("com.squareup.retrofit2:retrofit:2.9.0")
     implementation("com.squareup.retrofit2:converter-gson:2.9.0")
     implementation("com.google.code.gson:gson:2.10")
+    implementation ("io.insert-koin:koin-android:3.3.0")
```
Все новые зависимости стоит делать через version catalog механизм - файл `libs.versions.toml`, чтобы все внешние зависимости проекта находились в одном месте - так удобнее будет с ними работать, а также все модули проекта будут зависеть от одних и тех же версий сторонних зависимостей. В Android Studio можно это сделать автоматически - нужно поставить каретку на зависимость, нажать Alt + Enter (cmd + Enter на маке) и выбрать соответствующий пункт, чтобы автоматически мигрировать зависимость на versions catalog. Также в официальной документации есть гайд по миграции - https://developer.android.com/build/migrate-to-catalogs

ekalemen replied:
Спасибо за информацию, позже поизучаю как это делать

ArturNurtdinov commented on app/src/main/java/com/practicum/playlistmaker/sharing/data/ExternalNavigatorImpl.kt:
```diff
@@ -10,10 +10,12 @@ import com.practicum.playlistmaker.sharing.domain.model.EmailData
         shareIntent.setType("text/plain")
-        shareIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
-        context.startActivity(Intent.createChooser(shareIntent,"Share link"))
+        val choosenIntent = Intent.createChooser(shareIntent,"Share link")
```
Текст для запуска шторки с выбором приложения лучше перенести в strings.xml, т.к. этот текст тоже должен локализоваться

ekalemen replied:
Fixed

Commit:
Sprint#17 critical review fixes

This fixes:
 - move MediaPlayer creation to the Koin module
 - move Gson creation to the  Koin module
 - minor fixes

Signed-off-by: Eugeny Kalemenev <e.kalemenev@yadro.com>

RD2W commented on app/src/main/java/com/practicum/playlistmaker/search/di/SearchModule.kt:
```diff
@@ -0,0 +1,48 @@
+        TracksInteractorImpl(get())
+    }
+
+    viewModel {
```
Жень, привет!

Лучше использовать [constructor DSL](https://insert-koin.io/docs/quickstart/android#koin-module-classic-or-constructor-dsl), тогда получиться 
```kotlin
viewModelOf(::SearchViewModel)
```
и не нужно будет никаких get()

RD2W commented on app/src/main/java/com/practicum/playlistmaker/search/di/SearchModule.kt:
```diff
@@ -0,0 +1,48 @@
+        androidContext().getSharedPreferences("playlist_maker_prefs", Context.MODE_PRIVATE)
+    }
+
+    factory {
```
Аналогично используя constructor DSL

```kotlin
factoryOf(::Gson)
```

RD2W commented on app/src/main/java/com/practicum/playlistmaker/search/di/SearchModule.kt:
```diff
@@ -0,0 +1,48 @@
+import org.koin.dsl.module
+
+val SearchModule = module {
+    factory<SearchHistoryRepository> {
```
```kotlin
factoryOf(::SearchHistoryRepositoryImpl) { bind<SearchHistoryRepository>() }
```
